### ISSUE TITLE
Tab Grouping feature like in Microsoft Edge

### ISSUE THREAD

**Issue by symonyshen** [jupyterlab/jupyterlab#15245]
We are limited by the amount of tabs one can have open simultaneously in JupyterLab, not because of performance issues, but rather because we passed a threshold where we cannot see the name of the tabs in the actual GUI and the open tab section of the GUI gets messy and confusing. 

To counter this issue, to improve user experience and to boost productivity, a tab grouping mechanism could be implemented, following what Microsoft did with Edge in their last iteration of the browser.

This way we can create tab groupings, name them, open/close them, and add/remove a specific tab.

We believe this would also contribute to a better experience when coding through long sessions as one can organize their workspace better and cleaner.

Bellow you can find an example of what is currently implemented in Edge:

<img width="1358" alt="image" src="https://github.com/jupyterlab/jupyterlab/assets/134851942/d321641f-a4d6-43fd-acd8-46c5731f2928">

I am more than available to discuss this feature as we believe it would create value inside the platform.

Thank you very much.

Maksym Symonyshen

**Comment by krassowski** [comment]
It's great to hear that the number of tabs and not performance is the blocker!

I am not sure if tab grouping is something with a broad consensus for including by default (one can get a better judgement of this by attending a [weekly jupyterlab call](https://github.com/jupyterlab/team-compass#weekly-jupyterlab-team-meeting) and pitching this idea).
In either case I believe we would want to enable extension authors to make it possible via an extension - if you were to attempt to make an extension and find that public API does not permit it, please feel welcome to make issues with specifc suggestions and open pull requests.

Please also see related:
- https://github.com/jupyterlab/jupyterlab/issues/10305
- #12278 (I am not actively working on this PR but if you would like to revive this effort let me know)

**Comment by JasonWeill** [comment]
As with browser windows, a workaround for this is to open multiple JupyterLab windows; each window can have its own tabs. Accepting this in triage; if someone wants to code this as an extension, we would consider that pull request for inclusion. Thanks!

**Comment by GabrielaVives** [comment]
If the main concern is not being able to see the titles of the tabs when too many are open, should we consider adding a tooltip on hover to show the title? 
This is a standard for tab display, used in pretty much any context where the tab width can hide the title.

I don't see any issues with tab grouping, but it seems it won't solve the problem of not seeing the tab names, as they would be hidden in a group. I think users facing 2 separate issues here:
- being lost when tab titles are not visible, spending unnecessary time clicking and searching for a specific tab (solution could be: tooltips on hover)
- lacking clarity when having to work with many tabs open (solution could be: be able to organise them into meaningful groups)

**Comment by krassowski** [comment]
> If the main concern is not being able to see the titles of the tabs when too many are open, should we consider adding a tooltip on hover to show the title?

There already is a tooltip, and sadly it is far from a sufficient solution. If you have 30 tabs and tooltip takes 1 second to show and process, you lost half a minute looking for a tab, long enough to get frustrated.

**Comment by GabrielaVives** [comment]
> There already is a tooltip, and sadly it is far from a sufficient solution. If you have 30 tabs and tooltip takes 1 second to show and process, you lost half a minute looking for a tab, long enough to get frustrated.

Oh ok, I tried and didn't have a tooltip, sorry. 

Even if it doesn't solve the problem, 1 second seems a lot. If it's possible to have an instant tooltip when you hover over multiple tabs like on Chrome that would already be an upgrade, but I am not aware of the technical matters that would make this possible or not.

**Comment by JasonWeill** [comment]
If we're using DOM attributes like the `title` element, we don't have control over the lag associated with tooltips. We'd have to have our own tooltip UI component to display tooltips immediately. A tab list UI, or alternative tab displays such as vertical tabs, might also help.

**Comment by krassowski** [comment]
> If we're using DOM attributes like the title element, we don't have control over the lag associated with tooltips. We'd have to have our own tooltip UI component to display tooltips immediately. 

Yes, this is the case right now, and indeed having a custom tooltip could help.

> A tab list UI, or alternative tab displays such as vertical tabs, might also help.

We already have it but it is still not sufficient:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/2735be74-c4b9-4965-8320-02151309bfe9)

The current problems with the tab list I have are:
- a) I cannot have all three at once:
    - file browser
    - table of contents
    - tab list
- b) it does not highlight the active tab (harder to get a reference point)
- c) it does represent the spacial layout (for example on the screenshot above, I do not immediately see if the tab is from the  top or bottom panel)
- d) there is no context menu on the tab items in the tab list, so I still need to find the tab in the main area dock itself

**Comment by krassowski** [comment]
And there is the tab menu too:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/febce197-cdea-4fd1-92f4-fd4caa813737)

**Comment by krassowski** [comment]
A solution to the number of the limitations of the "Open Tabs" subpanel might be implementing https://github.com/jupyterlab/jupyterlab/issues/15293.
